packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts

```diff
@@ -311,39 +313,64 @@ export const createColumnsState = ({
   // TODO v6: remove the sync between the columns `hide` option and the model.
   let columnVisibilityModel: GridColumnVisibilityModel = {};
   if (shouldRegenColumnVisibilityModelFromColumns) {
-    if (reset) {
-      columnsStateWithPreProcessing.all.forEach((field) => {
-        columnVisibilityModel[field] = !columnsStateWithoutColumnVisibilityModel.lookup[field].hide;
-      });
-    } else {
-      const newColumnVisibilityModel = { ...currentColumnVisibilityModel };
-      let hasModelChanged = false;
+    let hasModelChanged = false;
 
+    const newColumnVisibilityModel = { ...currentColumnVisibilityModel };
+    if (isInsideStateInitializer) {
       columnsStateWithPreProcessing.all.forEach((field) => {
-        // If neither the `columnsToUpsert` nor the pre-processors updated the column,
-        // Then we don't want to update the visibility status of the column in the model.
-        if (
-          !columnsToUpsertLookup[field] &&
-          columnsLookupBeforePreProcessing[field] === columnsStateWithPreProcessing.lookup[field]
-        ) {
-          return;
+        newColumnVisibilityModel[field] =
+          !columnsStateWithoutColumnVisibilityModel.lookup[field].hide;
+      });
+    } else if (keepOnlyColumnsToUpsert) {
+      // At this point, `keepOnlyColumnsToUpsert` changes its meaning to keep the columns
```

Comment:
```suggestion
      // At this point, `keepOnlyColumnsToUpsert` have a new meaning: keep the columns
```

The sentence is valid but I had to read it a few times to understand

Reply:
Feel free to suggest improvements.